lib_eio_linux/eio_stubs.c

```diff
@@ -40,3 +45,37 @@ CAMLprim value caml_eio_getrandom(value v_ba, value v_off, value v_len) {
   if (ret == -1) uerror("getrandom", Nothing);
   CAMLreturn(Val_long(ret));
 }
+
+struct linux_dirent64 {
+    ino64_t        d_ino;    /* 64-bit inode number */
+    off64_t        d_off;    /* 64-bit offset to next structure */
+    unsigned short d_reclen; /* Size of this dirent */
+    unsigned char  d_type;   /* File type */
+    char           d_name[]; /* Filename (null-terminated) */
+};
+
+CAMLprim value caml_eio_getdents(value v_fd) {
+  CAMLparam1(v_fd);
+  CAMLlocal1(result);
+  char buf[DIRENT_BUF_SIZE];
+  struct linux_dirent64 *d;
+  int nread, pos;
+  caml_enter_blocking_section();
+  nread = syscall(SYS_getdents64, Int_val(v_fd), buf, DIRENT_BUF_SIZE);
+  caml_leave_blocking_section();
+  if (nread == -1) uerror("getdents", Nothing);
+
+  result = Val_int(0); /* The empty list */
+
+  for (pos = 0; pos < nread;) {
+    d = (struct linux_dirent64 *) (buf + pos);
+    CAMLlocal1 (r);
```

Comment:
Not sure, but I suspect doing `CAMLlocal1` inside a loop is a bad idea (it's adding `r` to `caml__roots_`).

Reply:
Oops yes! Thanks